api/internal/handler/service/service_test.go

```diff
@@ -18,274 +18,832 @@
 
-	//sleep
-	time.Sleep(time.Duration(100) * time.Millisecond)
+			h := Handler{serviceStore: serviceStore, upstreamStore: upstreamStore}
+			ctx := droplet.NewContext()
+			ctx.SetInput(tc.giveInput)
+			ret, err := h.Create(ctx)
+			assert.Equal(t, tc.getCalled, getCalled)
+			assert.Equal(t, tc.wantRet, ret)
+			assert.Equal(t, tc.wantErr, err)
+		})
+	}
+}
 
-	//get
-	input11 := &GetInput{}
-	input11.ID = "11"
-	ctx.SetInput(input11)
-	ret, err = handler.Get(ctx)
-	stored = ret.(*entity.Service)
-	assert.Nil(t, err)
-	assert.Equal(t, "11", stored.ID)
+func TestService_Update(t *testing.T) {
+	tests := []struct {
+		caseDesc      string
+		getCalled     bool
+		giveInput     *UpdateInput
+		giveErr       error
+		giveRet       interface{}
+		wantInput     *entity.Service
+		wantErr       error
+		wantRet       interface{}
+		upstreamInput string
+		upstreamRet   interface{}
+		upstreamErr   interface{}
+	}{
+		{
+			caseDesc:  "create success",
+			getCalled: true,
+			giveInput: &UpdateInput{
+				ID: "s1",
+				Service: entity.Service{
+					Name:       "s1",
+					UpstreamID: "u1",
+					Desc:       "test service",
+				},
+			},
+			wantInput: &entity.Service{
+				BaseInfo: entity.BaseInfo{
+					ID: "s1",
+				},
+				Name:       "s1",
+				UpstreamID: "u1",
+				Desc:       "test service",
+			},
+			upstreamInput: "u1",
+			upstreamRet: entity.Upstream{
+				BaseInfo: entity.BaseInfo{
+					ID: "u1",
+				},
+			},
+		},
+		{
+			caseDesc: "create failed, different id",
+			giveInput: &UpdateInput{
+				ID: "s1",
+				Service: entity.Service{
+					BaseInfo: entity.BaseInfo{
+						ID: "s2",
+					},
+					Name:       "s1",
+					UpstreamID: "u1",
+					Desc:       "test service",
+				},
+			},
+			wantRet: &data.SpecCodeResponse{StatusCode: http.StatusBadRequest},
+			wantErr: fmt.Errorf("ID on path (s1) doesn't match ID on body (s2)"),
+		},
+		{
+			caseDesc: "update failed, upstream not found",
+			giveInput: &UpdateInput{
+				ID: "s1",
+				Service: entity.Service{
+					Name:       "s1",
+					UpstreamID: "u1",
+					Desc:       "test service",
+				},
+			},
+			wantInput: &entity.Service{
+				Name:       "s1",
+				UpstreamID: "u1",
+				Desc:       "test service",
+			},
+			wantErr:       fmt.Errorf("upstream id: u1 not found"),
+			wantRet:       &data.SpecCodeResponse{StatusCode: http.StatusBadRequest},
+			upstreamInput: "u1",
+			upstreamErr:   data.ErrNotFound,
+		},
+		{
+			caseDesc: "update failed, upstream return error",
+			giveInput: &UpdateInput{
+				ID: "s1",
+				Service: entity.Service{
+					Name:       "s1",
+					UpstreamID: "u1",
+					Desc:       "test service",
+				},
+			},
+			wantInput: &entity.Service{
+				Name:       "s1",
+				UpstreamID: "u1",
+				Desc:       "test service",
+			},
+			wantErr:       fmt.Errorf("unknown error"),
+			wantRet:       &data.SpecCodeResponse{StatusCode: http.StatusBadRequest},
+			upstreamInput: "u1",
+			upstreamErr:   fmt.Errorf("unknown error"),
+		},
+		{
```

Comment:
we could add a test case for service not found.

Reply:
It's hard to test this.
Because we don't call `upstream.Get` in the `Update` function.
What we can test is that the third argument in the `Update` function is true.

Reply:
OK